pkg/apis/openstack/validation/infrastructure.go

```diff
@@ -74,66 +78,104 @@ func ValidateInfrastructureConfigUpdate(oldConfig, newConfig *api.Infrastructure
 	var (
-		allErrs                = field.ErrorList{}
-		validValues            []string
-		globalCandidate        *api.FloatingPool
-		globalCandidateScore   int
-		regionalCandidate      *api.FloatingPool
-		regionalCandidateScore int
-		regionalFound          bool
+		allErrs               = field.ErrorList{}
+		validValues           = sets.NewString()
+		additionalValidValues = sets.NewString()
 	)
 
-	for _, fp := range floatingPools {
-		// store the first non-regional image for fallback value if no floating pool for the given
-		// region was found
-		if fp.Region == nil && !regionalFound {
-			if match, score := utils.SimpleMatch(fp.Name, floatingPoolName); match {
-				if globalCandidate == nil || globalCandidateScore < score {
-					f := fp
-					globalCandidate = &f
-					globalCandidateScore = score
-				}
+	found := findFloatingPoolSub(floatingPools, &domain, &shootRegion, floatingPoolName, false, validValues, additionalValidValues)
+	if found != nil {
+		return found, allErrs
+	}
+
+	nonConstrainingOnly := len(validValues) > 0
+	validValuesRegion := sets.NewString()
+	validValuesDomain := sets.NewString()
+	foundRegion := findFloatingPoolSub(floatingPools, nil, &shootRegion, floatingPoolName, nonConstrainingOnly, validValuesRegion, additionalValidValues)
+	foundDomain := findFloatingPoolSub(floatingPools, &domain, nil, floatingPoolName, nonConstrainingOnly, validValuesDomain, additionalValidValues)
+	if foundRegion != nil && foundDomain != nil {
+		return foundRegion, allErrs
+	}
+	if foundRegion != nil && len(validValuesDomain) == 0 {
+		return foundRegion, allErrs
+	}
+	if foundDomain != nil && len(validValuesRegion) == 0 {
+		return foundDomain, allErrs
+	}
+	if len(validValuesRegion) != 0 && len(validValuesDomain) != 0 {
+		for value := range validValuesRegion {
+			if validValuesDomain.Has(value) {
+				validValues.Insert(value)
 			}
-			continue
 		}
+	} else if len(validValuesRegion) != 0 {
+		validValues = validValuesRegion
+	} else if len(validValuesDomain) != 0 {
+		validValues = validValuesDomain
+	}
 
-		// floating pool for the given region found
-		if fp.Region != nil && *fp.Region == shootRegion {
-			regionalFound = true
-			validValues = append(validValues, fp.Name)
+	nonConstrainingOnly = len(validValues) > 0
+	found = findFloatingPoolSub(floatingPools, nil, nil, floatingPoolName, nonConstrainingOnly, validValues, additionalValidValues)
+	if found != nil {
+		return found, allErrs
+	}
+
+	validValuesList := []string{}
+	for key := range validValues {
+		validValuesList = append(validValuesList, key)
+	}
+	for key := range additionalValidValues {
+		validValuesList = append(validValuesList, key)
+	}
+	sort.Strings(validValuesList)
+	allErrs = append(allErrs, field.NotSupported(fldPath, floatingPoolName, validValuesList))
+	return nil, allErrs
+}
+
+func findFloatingPoolSub(floatingPools []api.FloatingPool, domain, shootRegion *string, floatingPoolName string,
```

Comment:
Can we find a better function name? What is "sub"?

Reply:
renamed to findFloatingPoolCandidate